jax/collect_profile.py

```diff
@@ -0,0 +1,90 @@
+import os
+import pathlib
+import tempfile
+
+import jax
+try:
+  from tensorflow.python.profiler import profiler_v2 as profiler
+  from tensorflow.python.profiler import profiler_client
+except ImportError:
+  raise ImportError("This script requires `tensorflow` to be installed.")
+try:
+  from tensorboard_plugin_profile.convert import raw_to_tool_data as convert
+except ImportError:
+  raise ImportError(
+      "This script requires `tensorboard_plugin_profile` to be installed.")
+
+
+_DESCRIPTION = """
+To profile running JAX programs, you first need to start the profiler server
+in the program of interest. You can do this via
+`jax.profiler.start_server(<port>)`. Once the program is running and the
+profiler server has started, you can run `collect_profile` to trace the execution
+for a provided duration. The trace file will be dumped into a directory
+(determined by `--log_dir`) and by default, a Perfetto UI link will be generated
+to view the resulting trace.
+"""
+parser = argparse.ArgumentParser(description=_DESCRIPTION)
+parser.add_argument("--log_dir", default=None,
+                    help=("Directory to store log files. "
+                          "Uses a temporary directory if none provided."),
+                    type=str)
+parser.add_argument("port", help="Port to collect trace", type=int)
+parser.add_argument("duration_in_ms",
+                    help="Duration to collect trace in milliseconds", type=int)
+parser.add_argument("--no_perfetto_link",
+                    help="Disable creating a perfetto link", action="store_true")
+parser.add_argument("--host", default="127.0.0.1",
+                    help="Host to collect trace. Defaults to 127.0.0.1",
+                    type=str)
+parser.add_argument("--host_tracer_level", default=2,
+                    help="Profiler host tracer level", type=int)
+parser.add_argument("--device_tracer_level", default=1,
+                    help="Profiler device tracer level", type=int)
+parser.add_argument("--python_tracer_level", default=0,
+                    help="Profiler Python tracer level", type=int)
+
+def main(args):
+  options = profiler.ProfilerOptions(
+      host_tracer_level=args.host_tracer_level,
+      device_tracer_level=args.device_tracer_level,
+      python_tracer_level=args.python_tracer_level,
+  )
+  log_dir = pathlib.Path(args.log_dir if args.log_dir else tempfile.mkdtemp())
+  profiler_client.trace(
+      f"{args.host}:{args.port}",
+      str(log_dir),
+      args.duration_in_ms,
+      options=options)
+  print(f"Dumped profiling information in: {log_dir}")
+  curr_path = log_dir.resolve()
+  root_trace_folder = curr_path / "plugins" / "profile"
+  trace_folders = [root_trace_folder / trace_folder for trace_folder
+                   in root_trace_folder.iterdir()]
+  latest_folder = max(trace_folders, key=os.path.getmtime)
+  xplane = next(latest_folder.glob("*.xplane.pb"))
+  result = convert.xspace_to_tool_data([xplane], "trace_viewer^", None)
```

Comment:
What is this doing?

Reply:
The `profiler.trace(...)` outputs an `xplane.pb` file that needs to be converted into `trace.json` for use w/ Perfetto UI. I'll add a comment explaining what we are doing here.

Reply:
Done.